Jellyfin.Server.Implementations/UserData/UserDataManager.cs

```diff
@@ -0,0 +1,273 @@
+    {
+        ArgumentNullException.ThrowIfNull(userData);
+
+        ArgumentNullException.ThrowIfNull(item);
+
+        ArgumentNullException.ThrowIfNull(user);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var keys = item.GetUserDataKeys();
+
+        var dbContext = await _provider.CreateDbContextAsync(cancellationToken).ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            foreach (var key in keys)
+            {
+                dbContext.UserDatas.Update(userData);
+            }
+
+            await dbContext.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        UserDataSaved?.Invoke(this, new UserDataSaveEventArgs
+        {
+            Keys = keys,
+            UserData = userData,
+            SaveReason = reason,
+            UserId = user!.Id,
+            Item = item
+        });
+    }
+
+    public void SaveUserData(Guid userId, BaseItem item, Data.Entities.UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        var user = _userManager.GetUserById(userId);
+        SaveUserData(user, item, userData, reason, cancellationToken);
+    }
+
+    public void SaveUserData(User? user, BaseItem item, Data.Entities.UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        SaveUserDataAsync(user, item, userData, reason, cancellationToken).ConfigureAwait(true);
+    }
+
+    private async Task<UserItemData> GetUserDataAsync(User? user, BaseItem item)
+    {
+        ArgumentNullException.ThrowIfNull(user);
+        ArgumentNullException.ThrowIfNull(item);
+        var dbContext = await _provider.CreateDbContextAsync().ConfigureAwait(true);
+        UserItemData? userItemData;
+        await using (dbContext.ConfigureAwait(true))
+        {
+            userItemData = await dbContext.UserDatas.FindAsync(new object[] { user.Id, item.Id.ToString() }).ConfigureAwait(true) ?? await CreateUserDataInternalAsync(user, item, dbContext).ConfigureAwait(true);
+        }
+
+        return userItemData;
+    }
+
+    internal async Task<UserItemData> CreateUserDataInternalAsync(User user, BaseItem item, DbContext dbContext)
+    {
+        UserItemData newUserData = new UserItemData
+        {
+            UserId = user.Id,
+            Key = item.Id.ToString()
+        };
+
+        await dbContext.AddAsync(newUserData).ConfigureAwait(true);
+        await dbContext.SaveChangesAsync().ConfigureAwait(true);
+
+        return newUserData;
+    }
+
+    public UserItemData GetUserData(Guid userId, BaseItem item)
+    {
+        var user = _userManager.GetUserById(userId);
+
+        return GetUserData(user, item);
+    }
+
+    public UserItemData GetUserData(User? user, BaseItem item)
+    {
+        return GetUserDataAsync(user, item).GetAwaiter().GetResult();
```

Comment:
I would either deprecate the sync function or make a parallel sync `GetUserDataAsync` one

Reply:
Sory, that was maybe ambiguous.
What i meant, remove this method and update all callers to use the async instead, or make a totally Sync function.